### PR TITLE
DDB Mapper filter expressions (runtime components)

### PR DESCRIPTION
**Pull request by ianbotsf**
## Issue \#

Related to #76

## Description of changes

This change adds low-level filter expressions support to the runtime, although nothing uses them yet. The filter DSL is defined in the `Filter` interface and will _eventually_ (in the next PR) be part of a Lambda signature in a codegenned member for operation requests.

Something like this:

```kotlin
// Generated code
interface QueryRequest<T> {
    // Property syntax
    val filter: Filter.() -> BooleanExpr
    
    // DSL method syntax
    fun filter(block: Filter.() -> BooleanExpr)

    // …all other Query members
}
```

Callers will then use this when building their query structures to specify their filter expression (if desired):

```kotlin
// User code
table.query {
    filter {
        and(
            attr("productName") eq "Nimbus 2000",
            attr("condition") isIn setOf("New", "Refurbished"),
            attr("price") lt 500.00,
        )
    }

    // …set other Query members
}
```

In addition to the normal things, I could _especially_ use feedback on the DSL—names, conventions, types, etc. If it's incomprehensible to you then it'll probably be even worse for users!

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: hll/dynamodb-mapper/dynamodb-mapper/common/src/aws/sdk/kotlin/hll/dynamodbmapper/expressions/Filter.kt [hunk 1]
```diff
@@ -0,0 +1,920 @@
...
+ * # Other functions/properties
+ *
+ * Several additional filter expressions are possible via the following methods/properties:
+ *
+ * * [contains] — checks if a string/list attribute value contains the given value
+ * * [exists] — checks if _any value_ (including `null`) exists for an attribute
+ * * [notExists] — checks if no value is present for an attribute (i.e., the attribute is "undefined" for an item)
+ * * [isOfType] — checks if an attribute value is of the given type
+ * * [size] — gets the size of an attribute (e.g., number of elements in list/map/set, the length of a string, etc.)
+ * * [startsWith] — checks if a string attribute value starts with the given value
```
Comment (ianbotsf) [inline hll/dynamodb-mapper/dynamodb-mapper/common/src/aws/sdk/kotlin/hll/dynamodbmapper/expressions/Filter.kt:169]: **live review**: Here and elsewhere when referring to Kotlin-ized names of DDB primitives, mention the underlying primitive to help users get up to speed, transition from another mapper, etc.

Comment (ianbotsf) [reply hll/dynamodb-mapper/dynamodb-mapper/common/src/aws/sdk/kotlin/hll/dynamodbmapper/expressions/Filter.kt:169]: Also add a few examples here.

File: settings.gradle.kts [hunk 2]
```diff
@@ -75,7 +75,7 @@ if ("dynamodb".isBootstrappedService) {
     include(":hll:dynamodb-mapper:dynamodb-mapper-annotations")
     include(":hll:dynamodb-mapper:tests:dynamodb-mapper-annotation-processor-test")
     include(":hll:dynamodb-mapper:dynamodb-mapper-schema-generator-plugin")
-    include(":hll:dynamodb-mapper:tests:dynamodb-mapper-schema-generator-plugin-test")
+    // include(":hll:dynamodb-mapper:tests:dynamodb-mapper-schema-generator-plugin-test")
```
Comment (ianbotsf) [inline settings.gradle.kts:78]: **live review**: accidental inclusion, remove

File: hll/dynamodb-mapper/dynamodb-mapper/common/src/aws/sdk/kotlin/hll/dynamodbmapper/expressions/Expressions.kt [hunk 3]
```diff
@@ -0,0 +1,486 @@
...
+ */
+public sealed interface Expression {
+    /**
+     * Accepts a visitor that is traversing an expression tree by dispatching to a subtype implementation. Subtype
+     * implementations MUST call the [ExpressionVisitor.visit] overload for their concrete type (effectively forming a
+     * [double dispatch](https://en.wikipedia.org/wiki/Double_dispatch)) and MUST return the resulting value.
+     * @param visitor The [ExpressionVisitor] that is traversing an expression
+     */
+    public fun <T> accept(visitor: ExpressionVisitor<T>): T
+}
```
Comment (ianbotsf) [inline hll/dynamodb-mapper/dynamodb-mapper/common/src/aws/sdk/kotlin/hll/dynamodbmapper/expressions/Expressions.kt:31]: **live review**: split out hierarchy into different smaller files

**Review by ianbotsf — COMMENTED** [review]
Captured some notes from a live code review with the team today. More comments/reviews to follow.

File: hll/dynamodb-mapper/dynamodb-mapper/common/src/aws/sdk/kotlin/hll/dynamodbmapper/expressions/Comparator.kt [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+     */
+    LESS_THAN_OR_EQUAL("<="),
+
+    /**
+     * A greater-than comparison, equivalent to `>` in Kotlin and DynamoDB
+     */
+    GREATER_THAN(">"),
+
+    /**
+     * A greater-than-or-equal-to comparison, equivalent to `>=` in Kotlin
```
Comment (lauzadis) [inline hll/dynamodb-mapper/dynamodb-mapper/common/src/aws/sdk/kotlin/hll/dynamodbmapper/expressions/Comparator.kt:38]: nit: "in Kotlin and DynamoDB"?

File: hll/dynamodb-mapper/dynamodb-mapper/common/src/aws/sdk/kotlin/hll/dynamodbmapper/expressions/internal/ExpressionImpls.kt
Comment (lauzadis) [inline hll/dynamodb-mapper/dynamodb-mapper/common/src/aws/sdk/kotlin/hll/dynamodbmapper/expressions/internal/ExpressionImpls.kt:1]: opinion: I feel like the implementations don't need to repeat the KDocs of what they are implementing

**Review by lauzadis — APPROVED** [review]
No strong comments / suggestions outside of what we covered in the live review!

File: hll/dynamodb-mapper/dynamodb-mapper/common/src/aws/sdk/kotlin/hll/dynamodbmapper/expressions/Filter.kt [hunk 5]
```diff
@@ -0,0 +1,920 @@
...
+     * Creates an equality expression for verifying two expressions are equal to each other
+     * @param expr The other expression in the comparison
+     */
+    public infix fun Expression.eq(expr: Expression): BooleanExpr
+
+    /**
+     * Creates an equality expression for verifying two expressions are equal to each other
+     * @param value The other value in the comparison
+     */
+    public infix fun Expression.eq(value: Boolean?): BooleanExpr = eq(LiteralExpr(value))
```
[the remark this answers is not in the document]
Comment (ianbotsf) [reply hll/dynamodb-mapper/dynamodb-mapper/common/src/aws/sdk/kotlin/hll/dynamodbmapper/expressions/Filter.kt:205]: Turns out we can but it only affects Dokka-generated documentation. IntelliJ will only find KDocs directly attached to each overload. I'm going to leave the documentation on overloads for the time being.

File: hll/dynamodb-mapper/dynamodb-mapper/common/src/aws/sdk/kotlin/hll/dynamodbmapper/expressions/internal/ExpressionImpls.kt
Comment (ianbotsf) [reply hll/dynamodb-mapper/dynamodb-mapper/common/src/aws/sdk/kotlin/hll/dynamodbmapper/expressions/internal/ExpressionImpls.kt:1]: Can you clarify what you mean? This `ScalarFunc` enum doesn't implement a particular interface...it identifies a particular low-level function.

Comment (lauzadis) [reply hll/dynamodb-mapper/dynamodb-mapper/common/src/aws/sdk/kotlin/hll/dynamodbmapper/expressions/internal/ExpressionImpls.kt:1]: This comment is referring to the entire `ExpressionImpls.kt` file

Comment (lauzadis) [reply hll/dynamodb-mapper/dynamodb-mapper/common/src/aws/sdk/kotlin/hll/dynamodbmapper/expressions/internal/ExpressionImpls.kt:1]: Yeah, I was thinking in this case they are repeated on the public interfaces, so the internal implementations don't need them

**Review by 0marperez — APPROVED** [review]

**Review by lauzadis — APPROVED** [review]

**Pull request merged**